[WIP] Improve slash commands

neiljp opened the pull request:
So far this PR:
* adds typeahead for the existing `/ping` and `/todo` slash commands
* adjusts the typeahead to show prioritized commands
* adds the `/theme` typeahead as a generalized version of dark/light/night/day (it replaces /dark in the prioritized commands, but the others are still present)
* adds a tip to slash commands, to an unimplemented means to list all slash commands (not easily otherwise achieved)

I'd definitely like to add more tests, so just putting this out to get feedback on these ideas and implementation first.

Except for the last commit this is all fully implemented and should work.

neiljp committed:
slash commands: Add `/theme` to generalize existing commands.

This mimics the existing specific dark/night/light/day commands as theme
names.

The command is added to typeahead, replacing `/dark` in the top-5 since
it is a more general command.

neiljp committed:
typeahead: Prioritize selected slash commands.

Previously typeahead took the top 5 (typeahead default number) of slash
commands, with matches sorted. To improve discovery and prioritize
useful commands after only typing `/`, this commit removes the sorting
and relies upon separate manual ordering of the slash commands in two
groups.

Current top-5 prioritized commands are:
* /dark
* /me
* /poll
* /settings
* /todo

neiljp committed:
typeahead: Add `/ping` and `/todo` slash commands.

neiljp committed:
WIP: Add tip message suggesting `/commands` to receive list of all slash commands.

NOTE: This is a prototype and not yet implemented!

showell commented:
I support the first two here.  I am meh on theme.  The last commit is intriguing.

neiljp committed:
slash commands: Use 'Switch to' instead of 'Toggle' in descriptions.

Running `/dark` while in dark/night mode does not switch to light/day
mode, so using the term 'toggle' is incorrect; this applies to all these
slash commands, as well as those adjusting the width setting.

neiljp commented:
@showell Rebased to put description adjustment at top to enable merging of early commits more easily.

neiljp committed:
slash commands: Allow and ignore additional slash-command parameters.

Prior to this, commands such as `/ping blah` would be posted as messages.

neiljp committed:
slash commands: Avoid accidental sending of empty `/me` messages.

neiljp commented:
@showell Added a few more commits - how do you feel about the first five? Should we add more tests too?

showell commented:
The first three seem fine to me, and it might make sense to just push them in their own PR to get a clean build.

For the validation piece, it's not clear to me how we want to handle things like "/day job" or "/dark knight".  On master we send those as messages, which does seem like the worst solution.  In your PR we would treat those like "/day" and "dark" and ignore the extra stuff, which is better.  But I wonder if we should just make those an error.

I would like to do the validation somewhat holistically, where we tackle it basically all at once.  I would think a nice paradigm is that every slash command (including things that are technically message directives like "me" and "poll")  would have a function that returns one of several possible values:

* skip - means it's not relevant syntax
* deliver - means we should send it as a message (and the caller basically continues with the normal message-send codepath)
* done - means the slash command was executed (and then the caller just clears the compose box)
* error - means there was a syntax error (and we should show the error and prevent delivery till the user overrides it)

And if we get "skip" returned from all slash commands, that means we have something like "/etc/host needs to be updated", in which case we should probably have the user confirm the message before we deliver it.

I'm a little wary to invest too much into this if we decide it's better to just move away from slash commands, and Tim seems to be leaning in the direction of something more like alt-/ or something.

showell commented:
@neiljp Just a quick reminder that you can probably create a new PR with the first three commits here to get a clean build, and we can at least get those merged, since I don't expect any pushback on the first three.
